### PR TITLE
:new: :mortar_board: Topic12 `ArrayQueue` and `ArrayQueueTest` code added to repo

### PR DESCRIPTION
**Pull request by jameshughes89**
### What
Add the implementation and tests for `ArrayQueue`

### Where
src/main/java/ArrayQueue.java
src/test/java/ArrayQueueTest.java

### Testing
Ran tests locally. 

### Additional Notes
Comments are a little dry. Perhaps add more?

### CODE REVIEW

File: src/main/java/ArrayQueue.java [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+     */
+    @SuppressWarnings("unchecked")
+    private void expandCapacity() {
+        T[] newStack = (T[]) new Object[queue.length * 2];
+        for (int i = 0; i < queue.length; ++i) {
+            newStack[i] = queue[i];
+        }
+        queue = newStack;
+
+    }
```
Comment (twentylemon) [inline src/main/java/ArrayQueue.java:47]: Add some more tests about this behaviour. Where is the new element going to be inserted?

**Review by twentylemon — CHANGES_REQUESTED** [review]

File: src/main/java/ArrayQueue.java (hunk 1 above)
Comment (jameshughes89) [reply src/main/java/ArrayQueue.java:47]: bbbuuussssted

File: src/main/java/ArrayQueue.java [hunk 2]
```diff
@@ -35,15 +35,20 @@ public void enqueue(T element) {
...
         for (int i = 0; i < queue.length; ++i) {
-            newStack[i] = queue[i];
+            newQueue[i] = queue[front];
+            front = (front + 1) % queue.length;
         }
-        queue = newStack;
-
+        front = 0;
+        rear = size;
+        queue = newQueue;
```
Comment (twentylemon) [inline src/main/java/ArrayQueue.java:51]: You the fixed the bug here I think, but your tests never actually exposed it. I suggest reverting this and writing a new test that should pass, bu actually fails. Then, once you have exposed the bug, reapply these changes and see if that causes the test to pass (while still passing all the other tests, of course).

**Review by twentylemon — CHANGES_REQUESTED** [review]

**Comment by jameshughes89** [comment]
sad

**Comment by jameshughes89** [comment]
huzzah

**Comment by twentylemon** [comment]
> huzzah

The actual issue still isn't tested. I'll be explicit now since it's been two revisions now.

Assuming your old code where `front` and `back` are not updated, you don't have a test where you enqueue, dequeue then enqueue until capacity is full. After the dequeue, `front = back = 1`. During `expandCapacity` after all the enqueues you'll be at `front = 1, back = 0`. Going back to my original question,

> Where is the new element going to be inserted?

**Comment by jameshughes89** [comment]
> > huzzah
> 
> The actual issue still isn't tested. I'll be explicit now since it's been two revisions now.
> 
> Assuming your old code where `front` and `back` are not updated, you don't have a test where you enqueue, dequeue then enqueue until capacity is full. After the dequeue, `front = back = 1`. During `expandCapacity` after all the enqueues you'll be at `front = 1, back = 0`. Going back to my original question,
> 
> > Where is the new element going to be inserted?

Based on how the `enqueue` & `dequeue` are written, when `expandCapacity` is called, `front = back = 1` still, but that is beside the point and I believe I see what you're saying:

```java
    @Test
    @DisplayName("Enqueuing 6 elements into the queue expands capacity.")
    void enqueuingBeyondCapacityCallsExpandCapacityToMakeRoom() {
        Queue<Integer> queue = new ArrayQueue<>(5);
        queue.enqueue(99);
        queue.dequeue();
        // front = rear = 1
        for (int i = 0; i < 6; ++i) {
            queue.enqueue(i);
        }
        // After 6 enqueues, expandCapacity is called
        // Ensure first was not overwritten by 6th
        // consecutive enqueue
        assertEquals(0, queue.first(), "First was overwritten.");
        for (int i = 1; i < 5; ++i) {
            queue.dequeue();
        }
        // After 5 dequeues, the 6th element will be first
        assertEquals(5, queue.first());
    }
```

**Comment by jameshughes89** [comment]
> > > huzzah
> > 
> > 
> > The actual issue still isn't tested. I'll be explicit now since it's been two revisions now.
> > Assuming your old code where `front` and `back` are not updated, you don't have a test where you enqueue, dequeue then enqueue until capacity is full. After the dequeue, `front = back = 1`. During `expandCapacity` after all the enqueues you'll be at `front = 1, back = 0`. Going back to my original question,
> > > Where is the new element going to be inserted?
> 
> Based on how the `enqueue` & `dequeue` are written, when `expandCapacity` is called, `front = back = 1` still, but that is beside the point and I believe I see what you're saying:
> 
> ```java
>     @Test
>     @DisplayName("Enqueuing 6 elements into the queue expands capacity.")
>     void enqueuingBeyondCapacityCallsExpandCapacityToMakeRoom() {
>         Queue<Integer> queue = new ArrayQueue<>(5);
>         queue.enqueue(99);
>         queue.dequeue();
>         // front = rear = 1
>         for (int i = 0; i < 6; ++i) {
>             queue.enqueue(i);
>         }
>         // After 6 enqueues, expandCapacity is called
>         // Ensure first was not overwritten by 6th
>         // consecutive enqueue
>         assertEquals(0, queue.first(), "First was overwritten.");
>         for (int i = 0; i < 5; ++i) {
>             queue.dequeue();
>         }
>         // After 5 dequeues, the 6th element will be first
>         assertEquals(5, queue.first());
>     }
> ```

End up doing fill FIFO check here.

File: src/test/java/ArrayQueueTest.java [hunk 3]
```diff
@@ -47,27 +47,31 @@ void enqueuingOneItemIsFirst() {
...
-        // After 5 dequeues, the 6th element will be first
-        assertEquals(5, queue.first());
     }
 
     @Test
-    @DisplayName("Enqueueing and Dequeuing returns in FIFO order.")
-    void enqueueingAndDequeuingReturnsElementsInFIFOOrder() {
-        Queue<Integer> queue = new ArrayQueue<>();
+    @DisplayName("Enqueuing 6 elements into the queue expands capacity.")
+    void enqueuingBeyondCapacityCallsExpandCapacityToMakeRoom() {
```
Comment (twentylemon) [inline src/test/java/ArrayQueueTest.java:63]: Names, I'd include something like "while maintaining queue order" so it's clear that it's not just a simple capacity check.

**Review by twentylemon — APPROVED** [review]

**Pull request merged**